token/core/common/wallet.go

```diff
@@ -0,0 +1,225 @@
+		DeserializerProvider: deserializerProvider,
+		WalletFactory:        walletFactory,
+		Registries:           registries,
+	}
+}
+
+func (s *WalletService) RegisterOwnerWallet(id string, path string) error {
+	return s.Registries[driver.OwnerRole].Registry.RegisterIdentity(id, path)
+}
+
+func (s *WalletService) RegisterIssuerWallet(id string, path string) error {
+	return s.Registries[driver.IssuerRole].Registry.RegisterIdentity(id, path)
+}
+
+func (s *WalletService) GetAuditInfo(id view.Identity) ([]byte, error) {
+	return s.IdentityProvider.GetAuditInfo(id)
+}
+
+func (s *WalletService) GetEnrollmentID(auditInfo []byte) (string, error) {
+	return s.IdentityProvider.GetEnrollmentID(auditInfo)
+}
+
+func (s *WalletService) GetRevocationHandler(auditInfo []byte) (string, error) {
+	return s.IdentityProvider.GetRevocationHandler(auditInfo)
+}
+
+func (s *WalletService) RegisterRecipientIdentity(data *driver.RecipientData) error {
+	if data == nil {
+		return errors.Errorf("cannot recigest empty recipient")
+	}
+	s.Logger.Debugf("register recipient identity [%s] with audit info [%s]", data.Identity.String(), hash.Hashable(data.AuditInfo).String())
+
+	// recognize identity and register it
+	d, err := s.DeserializerProvider()
+	if err != nil {
+		return errors.Wrap(err, "failed getting deserializer")
+	}
+
+	// match identity and audit info
+	err = d.Match(data.Identity, data.AuditInfo)
+	if err != nil {
+		return errors.Wrapf(err, "failed to match identity to audit infor for [%s:%s]", data.Identity, hash.Hashable(data.AuditInfo))
+	}
+
+	// register verifier and audit info
+	v, err := d.GetOwnerVerifier(data.Identity)
+	if err != nil {
+		return errors.Wrapf(err, "failed getting verifier for [%s]", data.Identity)
+	}
+	if err := s.IdentityProvider.RegisterVerifier(data.Identity, v); err != nil {
+		return errors.Wrapf(err, "failed registering verifier for [%s]", data.Identity)
+	}
+	if err := s.IdentityProvider.RegisterAuditInfo(data.Identity, data.AuditInfo); err != nil {
+		return errors.Wrapf(err, "failed registering audit info for [%s]", data.Identity)
+	}
+
+	return nil
+}
+
+func (s *WalletService) Wallet(identity view.Identity) driver.Wallet {
+	w, _ := s.OwnerWallet(identity)
+	if w != nil {
+		return w
+	}
+	iw, _ := s.IssuerWallet(identity)
+	if iw != nil {
+		return iw
+	}
+	return nil
+}
+
+func (s *WalletService) OwnerWalletIDs() ([]string, error) {
+	return s.Registries[driver.OwnerRole].Registry.WalletIDs()
+}
+
+func (s *WalletService) OwnerWallet(id any) (driver.OwnerWallet, error) {
+	w, err := s.walletByID(driver.OwnerRole, id)
+	if err != nil {
+		return nil, err
+	}
+	return w.(driver.OwnerWallet), nil
+}
+
+func (s *WalletService) IssuerWallet(id any) (driver.IssuerWallet, error) {
+	w, err := s.walletByID(driver.IssuerRole, id)
+	if err != nil {
+		return nil, err
+	}
+	return w.(driver.IssuerWallet), nil
+}
+
+func (s *WalletService) AuditorWallet(id any) (driver.AuditorWallet, error) {
```

Comment:
Not really important but you could skip the repetition of these lines if you change the signature of `walletByID`:

```
func walletByID[T driver.Wallet](entry RegistryEntry, id any) (T, error)
```
Then all your methods above would become like this:
```
func (s *WalletService) AuditorWallet(id any) (driver.AuditorWallet, error) {
 return walletByID[driver.AuditorWallet](s.Registries[driver.AuditorRole], id)
}
```

Reply:
but then I have to pass also `s.WalletFactory`. I thought about it, but it gets ugly in other ways. No?

Reply:
aaah sorry missed that one!